hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,563 @@
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link
+ * HederaFunctionality#TOKEN_AIRDROP}.
+ */
+@Singleton
+public class TokenAirdropHandler implements TransactionHandler {
+
+    private final CryptoTransferValidator validator;
+
+    /**
+     * Default constructor for injection.
+     */
+    @Inject
+    public TokenAirdropHandler(@NonNull final CryptoTransferValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        pureChecks(context.body());
+
+        final var op = context.body().tokenAirdropOrThrow();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+        final var tokenStore = context.createStore(ReadableTokenStore.class);
+
+        for (final var transfers : op.tokenTransfers()) {
+            final var tokenID = transfers.tokenOrThrow();
+            final var tokenMeta = tokenStore.getTokenMeta(transfers.tokenOrElse(TokenID.DEFAULT));
+            validateTruePreCheck(tokenMeta != null, INVALID_TOKEN_ID);
+            checkFungibleTokenTransfers(tokenID, transfers.transfers(), context, accountStore);
+            checkNftTransfers(tokenID, transfers.nftTransfers(), context, tokenMeta, accountStore);
+        }
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+        final var op = txn.tokenAirdropOrThrow();
+        validator.airdropsPureChecks(op);
+    }
+
+    @Override
+    public void handle(@NonNull final HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenAirdropOrThrow();
+        final var pendingStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        var recordBuilder = context.recordBuilders().getOrCreate(TokenAirdropRecordBuilder.class);
+        List<TokenTransferList> tokenTransferListList = new ArrayList<>();
+
+        // charge custom fees in advance
+        var convertedOp = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(op.tokenTransfers())
+                .build();
+        assessCustomFee(context, convertedOp);
+
+        for (final var xfers : op.tokenTransfers()) {
```

Comment:
We should probably extract this logic to smaller methods, that do 
1. get lists for fungible transfers that can be transferred directly , and which can be pending airdrops
2.  get lists for nft transfers that can be transferred directly , and which can be pending airdrops
3. Construct a `CryptoTransfer` and dispatch with transferrable items
4. Put pending airdrop objects to state. 

That will clarify the same implementation with small restructure. What do you think ?